pkg/providers/vmware/provider.go

```diff
@@ -0,0 +1,422 @@
+	}, nil
+}
+
+// Close is a no-op which is present in order to satisfy the Provider interface.
+func (r *VmwareProvider) Close() {
+	// nothing to do
+}
+
+// ValidateDiskStatus is a no-op which is present in order to satisfy the Provider interface.
+func (r *VmwareProvider) ValidateDiskStatus(_ string) (bool, error) {
+	return true, nil
+}
+
+func (r *VmwareProvider) getClient() (*vclient.RichVmwareClient, error) {
+	if r.vmwareClient == nil {
+		c, err := r.factory.NewVmwareClient(r.vmwareSecretDataMap)
+		if err != nil {
+			return nil, err
+		}
+		r.vmwareClient = c.(*vclient.RichVmwareClient)
+	}
+	return r.vmwareClient, nil
+}
+
+func (r *VmwareProvider) getVM() (*object.VirtualMachine, error) {
+	if r.vm == nil {
+		err := r.LoadVM(r.instance.Spec.Source)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return r.vm, nil
+}
+
+func (r *VmwareProvider) getVmProperties() (*mo.VirtualMachine, error) {
+	vmwareClient, err := r.getClient()
+	if err != nil {
+		return nil, err
+	}
+	vm, err := r.getVM()
+	if err != nil {
+		return nil, err
+	}
+	if r.vmProperties == nil {
+		properties, err := vmwareClient.GetVMProperties(vm)
+		if err != nil {
+			return nil, err
+		}
+		r.vmProperties = properties
+	}
+	return r.vmProperties, nil
+}
+
+func (r *VmwareProvider) prepareDataVolumeCredentials() (*mapper.DataVolumeCredentials, error) {
+	username := r.vmwareSecretDataMap[usernameKey]
+	password := r.vmwareSecretDataMap[passwordKey]
+
+	secret, err := r.ensureSecretIsPresent(username, password)
+	if err != nil {
+		return &mapper.DataVolumeCredentials{}, err
+	}
+
+	return &mapper.DataVolumeCredentials{
+		URL:        r.vmwareSecretDataMap[apiUrlKey],
+		Thumbprint: r.vmwareSecretDataMap[thumbprintKey],
+		Username:   username,
+		Password:   password,
+		SecretName: secret.Name,
+	}, nil
+}
+
+func (r *VmwareProvider) ensureSecretIsPresent(keyAccess, keySecret string) (*corev1.Secret, error) {
+	vmiName := k8stypes.NamespacedName{
+		Name:      r.vmiObjectMeta.Name,
+		Namespace: r.vmiObjectMeta.Namespace,
+	}
+	secret, err := r.secretsManager.FindFor(vmiName)
+	if err != nil {
+		return nil, err
+	}
+	if secret == nil {
+		secret, err = r.createSecret(keyAccess, keySecret)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return secret, nil
+}
+
+func (r *VmwareProvider) createSecret(username, password string) (*corev1.Secret, error) {
+	vmiName := k8stypes.NamespacedName{
+		Name:      r.vmiObjectMeta.Name,
+		Namespace: r.vmiObjectMeta.Namespace,
+	}
+	newSecret := corev1.Secret{
+		Data: map[string][]byte{
+			keyAccessKey: []byte(username),
+			keySecretKey: []byte(password),
+		},
+	}
+	newSecret.OwnerReferences = []metav1.OwnerReference{
+		ownerreferences.NewVMImportOwnerReference(r.vmiTypeMeta, r.vmiObjectMeta),
+	}
+	err := r.secretsManager.CreateFor(&newSecret, vmiName)
+	if err != nil {
+		return nil, err
+	}
+	return &newSecret, nil
+}
+
+func foldErrors(errs []error, vmiName k8stypes.NamespacedName) error {
```

Comment:
Seems like a good candidate for moving to utils package; ovirt provider defines exactly the same function.